### PR TITLE
[SPARK-8484] [ML]. Added TrainValidationSplit for hyper-parameter tuning.

### PR DESCRIPTION
**Pull request by zapletal-martin**
- [X] Added TrainValidationSplit for hyper-parameter tuning. It randomly splits the input dataset into train and validation and use evaluation metric on the validation set to select the best model. It should be similar to CrossValidator, but simpler and less expensive.
- [X] Simplified replacement of https://github.com/apache/spark/pull/6996

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  def getTrainRatio: Double = $(trainRatio)
+
+  setDefault(trainRatio -> 0.75)
+}
+
+/**
+ * :: Experimental ::
+ * Validation for hyper-parameter tuning.
+ * Randomly splits the input dataset into train and validation sets.
+ * And uses evaluation metric on the validation set to select the best model.
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: nit: "...validation sets, and uses ..." (comma instead of period)

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ * Similar to CrossValidator, but only splits the set once.
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: Surround with braces like "[[CrossValidator]]" so API docs generate a link

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ */
+@Experimental
+class TrainValidatorSplit(override val uid: String) extends Estimator[TrainValidatorSplitModel]
+  with TrainValidatorSplitParams with Logging {
+
+  def this() = this(Identifiable.randomUID("cv"))
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: How about `"tcv"` (since `"cv"` is already used by CrossValidator)

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  /** @group setParam */
+  def setEstimatorParamMaps(value: Array[ParamMap]): this.type = set(estimatorParamMaps, value)
+
+  /** @group setParam */
+  def setEvaluator(value: Evaluator): this.type = set(evaluator, value)
+
+  /** @group setParam */
+  def setTrainRatio(value: Double): this.type = set(trainRatio, value)
+
+  private[this] def sample[T: ClassTag](
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: Is `ClassTag` required here? I'm uncertain...

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      rdd: RDD[T],
+      lb: Double,
+      ub: Double,
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: Is `ub` = 1 in all the usages? In which case, I think just providing a single `ratio` argument would be clearer.

**Comment by feynmanliang** [comment]
Done for now. This is in very good shape.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala (hunk 4 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: Ignore that comment, tried running the code and saw that `PartitionwiseSampledRDD`needs it

**Comment by feynmanliang** [comment]
Actually, is there much difference between this and `CrossValidator` other than being able to specify `trainRatio`? There seems to be a lot of code duplication between the two.

**Comment by zapletal-martin** [comment]
@feynmanliang yes there is a lot of duplicated code. I attempted to refactor that slightly under https://github.com/apache/spark/pull/6996. 

It is an interesting idea to just call CrossValidator instead of implementing the logic. I will have a look into that. But I assume we still need the Params and Model specific for TrainValidationSplit? We also need to decide if we want to do that as part of this review or separately.

cc @mengxr

**Comment by zapletal-martin** [comment]
Both CrossValidator and TrainValidationSplit use sampling to split the data to training and validation. 

Currently CrossValidator does
* numFolds = 1 - not valid
* numFolds = 2 - 0.0 to 0.5 training, 0.5 to 1 validation and 0.0 to 0.5 validation and 0.5 to 1 training

TrainValidationSplit does
* 0.0 to trainRatio training, trainRatio to 1 validation

Therefore the logic is different and using TrainValidationSplit is not the same as just calling CrossValidator. Please let me know if the logic implemented by TrainValidationSplit is what was expected. We can then potentially address the code duplication.

**Comment by feynmanliang** [comment]
`CrossValidator` currently doesn't take `numFolds = 1` because it uses `MLUtils#kfolds` which does splits using `trainRatio = 1 - 1/numFolds`, which in this case would mean nothing goes into the training set.

If we give a `trainRatio` param to `CrossValidator` and modify `MLUtils#kfold` accordingly, then each fold could be split using your set `trainRatio` so setting `numFolds = 1, trainRatio = x` would achieve the same functionality as `TrainValidationSplit` with `trainRatio = x`.

Note that in this case we would have to change the default `trainRatio = 1 - 1/numFolds` behavior for the special case `numFolds = 1` to avoid the error case described in my first paragraph.

**Comment by feynmanliang** [comment]
Sorry, didn't address your questions.
 * I'm proposing to add a `trainRatio` param to `CrossValidatorParams`, eliminating the need for the `TrainValidatorSplit` Params and Model
 * The logic in `TrainValidationSplit` looks fine to me and is what I expected

**Comment by zapletal-martin** [comment]
Thanks @feynmanliang. As I mentioned I tried to address the code duplication in my previous PR differently than you propose, but we decided to go with the simplest option for now.

I agree what you are proposing makes sense. The only thing that worries me would be unclear purpose of trainRatio when numFolds != 1. In that case CrossValidator splits the dataset to numFolds subsets of the same size and the ratio of training and validation sets is given (e.g. with numFolds set to 4 the training set is 0.75 and validation is 0.25) and therefore the trainRatio param would not be used?

We could do that, document that approach and essentially get rid of TrainValidatorSplit or the other option would be to preserve TrainValidatorSplit as a wrapper around the functionality to avoid confusion of CrossValidator having the trainRatio param.

**Comment by feynmanliang** [comment]
Ah, I overlooked your point about when `1 - trainRatio > 1/numFolds` we will have overlapping folds. Also, I realized that what I'm proposing (adding a `trainRatio` for each fold) contradicts what [Wikipedia](https://en.wikipedia.org/wiki/Cross-validation_(statistics)#k-fold_cross-validation) defines for k-fold validation. Thanks for pushing back on this.

I agree with you that `TrainValidatorSplit` and `CrossValidator` have two different functionality and should be separate classes. I like the idea of both wrapping a common (perhaps more confusing) implementation with both `numFolds` and `trainRatio`; it differentiates the concepts in the public API but shares code in the implementation.

**Comment by zapletal-martin** [comment]
I think we need a decision how to approach this. I would prefer to focus on the public api and avoid the refactor in this review and then address that in another review as discussed in https://github.com/apache/spark/pull/6996.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+import org.apache.spark.rdd.{RDD, PartitionwiseSampledRDD}
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.Utils
+import org.apache.spark.util.random.BernoulliCellSampler
+
+/**
+ * Params for [[TrainValidatorSplit]] and [[TrainValidatorSplitModel]].
+ */
+private[ml] trait TrainValidatorSplitParams extends ValidatorParams {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: `TrainValidatorSplit` -> `TrainValidationSplit`

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  def setEvaluator(value: Evaluator): this.type = set(evaluator, value)
+
+  /** @group setParam */
+  def setTrainRatio(value: Double): this.type = set(trainRatio, value)
+
+  private[this] def sample[T: ClassTag](
+      rdd: RDD[T],
+      lb: Double,
+      ub: Double,
+      seed: Int = Utils.random.nextInt()): (RDD[T], RDD[T]) = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: Should the method be a one-liner: `val (train, validation) = df.randomSplit([trainRatio, 1 - trainRatio], seed)`?

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    copied
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Model from train validation split.
+ */
+@Experimental
+class TrainValidatorSplitModel private[ml] (
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: Missing parameter doc.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    override val uid: String,
+    val bestModel: Model[_],
+    val avgMetrics: Array[Double])
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: We don't have averaging. `avgMetrics` -> `validationMetrics`?

**Comment by mengxr** [comment]
@feynmanliang @zapletal-martin The changes in this PR look good to me except a few minor comments. As discussed in #6996, let's focus on the public API to get this merged first.

We can have another PR for code reuse. There would be more discussion, e.g., having a base class handling arbitrary slicing of the input data and making `CrossValidator` and `TrainValidationSplit` extend it. I actually think there will be more lines of code if we implement it that way. It might not be worth the trade-off.

**Comment by mengxr** [comment]
LGTM. Merged into master. Please create JIRAs for follow-up work, e.g, Python API, user guide, and refactoring (if it is useful). Thanks!
